FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/viewmodel/EkmLoginViewModel.kt

```diff
@@ -0,0 +1,160 @@
+                  )
+                  return@launch
+                }
+
+                if (!domainRulesResult.data.orgRules.usesKeyManager() ||
+                  !domainRulesResult.data.orgRules.mustAutoImportOrAutoGenPrvWithKeyManager()
+                ) {
+                  ekmLiveData.value = Result.exception(
+                    EkmNotSupportedException(domainRulesResult.data.orgRules)
+                  )
+                  return@launch
+                }
+
+                val notSupportedCombination =
+                  checkNotSupportedOrgRulesCombination(domainRulesResult.data.orgRules)
+
+                if (notSupportedCombination.isNotEmpty()) {
+                  ekmLiveData.value = Result.exception(
+                    OrgRulesCombinationNotSupportedException(
+                      orgRules = domainRulesResult.data.orgRules,
+                      combination = notSupportedCombination
+                    )
+                  )
+                  return@launch
+                }
+
+                val ekmPrivateResult = repository.getPrivateKeysViaEkm(
+                  context = context,
+                  ekmUrl = domainRulesResult.data.orgRules.keyManagerUrl
+                    ?: throw java.lang.IllegalArgumentException("key_manager_url is empty"),
+                  tokenId = tokenId
+                )
+
+                ekmLiveData.value = ekmPrivateResult
+              } else {
+                ekmLiveData.value = domainRulesResult
+              }
+            } else {
+              ekmLiveData.value = Result.error(
+                LoginResponse(
+                  ApiError(
+                    -1,
+                    context.getString(R.string.user_not_verified)
+                  ), false
+                )
+              )
+            }
+          }
+
+          else -> {
+            ekmLiveData.value = Result.exception(IllegalStateException("Unhandled error"))
+          }
+        }
+      } catch (e: Exception) {
+        ekmLiveData.value = Result.exception(e)
+      }
+    }
+  }
+
+  private fun checkNotSupportedOrgRulesCombination(orgRules: OrgRules):
+      Map<OrgRules.DomainRule, Boolean> {
+    val notSupportedCombination = mutableMapOf(
+      OrgRules.DomainRule.PRV_AUTOIMPORT_OR_AUTOGEN to true
+    )
+
+    when {
+      orgRules.mustAutoGenPassPhraseQuietly() -> {
+        notSupportedCombination[OrgRules.DomainRule.PASS_PHRASE_QUIET_AUTOGEN] = true
+      }
+
+      !orgRules.forbidStoringPassPhrase() -> {
+        notSupportedCombination[OrgRules.DomainRule.FORBID_STORING_PASS_PHRASE] = false
+      }
+
+      orgRules.mustSubmitToAttester() -> {
+        notSupportedCombination[OrgRules.DomainRule.ENFORCE_ATTESTER_SUBMIT] = true
+      }
+
+      !orgRules.forbidCreatingPrivateKey() -> {
+        notSupportedCombination[OrgRules.DomainRule.NO_PRV_CREATE] = false
+      }
+
+      else -> notSupportedCombination.clear()
+    }
+    return notSupportedCombination
+  }
+}
```

Comment:
Actually, this code is a little too dynamic for my taste :-) I took some serious effort to understand how it works - and I've already known what it's supposed to do.

I wish the code was more dumb. Like this:

```kt
if(orgRules.xxx == yy && orgRules.zz == ww) {
  throw UnsupportedOrgRulesException('xxx=yy,zz=ww')
}
if(!orgRules.xxx == ... ) {
  ...
}
...
```

Reply:
I've used that logic to simplify (simplify? haha) displaying errors. Currently, I create a map of unsupported combinations and print it for a user. It looks like this one

![image](https://user-images.githubusercontent.com/2863246/123972751-d0364600-d9c3-11eb-9593-3e9be943c5ab.png)

In that case, the following code shows the above dialog
```kotlin

is OrgRulesCombinationNotSupportedException -> {
                showInfoDialog(
                  dialogTitle = "",
                  dialogMsg = getString(
                    R.string.combination_of_org_rules_is_not_supported,
                    it.exception.combination
                  ),
                  isCancelable = true
                )
              }
```

Reply:
I know, I understood that during the review, but I really wish the code could be less clever / more basic.